daft/udf.py

```diff
@@ -70,79 +69,101 @@ def arg_keys(self) -> list[str]:
-        ), "Computed series must map 1:1 to the expressions that were evaluated"
-        function_parameter_name_to_index = {name: i for i, name in enumerate(expressions)}
-
-        args = []
-        for name in arg_keys:
-            # special-case to skip `self` since that would be a redundant argument in a method call to a class-UDF
-            if name == "self":
-                continue
-
-            assert name in pyvalues or name in function_parameter_name_to_index
-            if name in pyvalues:
-                args.append(pyvalues[name])
-            else:
-                args.append(evaluated_expressions[function_parameter_name_to_index[name]])
-
-        kwargs = {}
-        for name in kwarg_keys:
-            assert name in pyvalues or name in function_parameter_name_to_index
-            if name in pyvalues:
-                kwargs[name] = pyvalues[name]
-            else:
-                kwargs[name] = evaluated_expressions[function_parameter_name_to_index[name]]
-
-        # NOTE: We currently initialize the function once for every invocation of the PartialUDF.
-        # This is not ideal and we should cache initializations across calls for the same process.
-        func = self.udf.get_initialized_func()
-
-        try:
-            result = func(*args, **kwargs)
-        except Exception as user_function_exception:
-            raise RuntimeError(
-                f"User-defined function `{func.__name__}` failed when executing on inputs with lengths: {tuple(len(series) for series in evaluated_expressions)} error: {user_function_exception}"
-            ) from user_function_exception
-
-        # HACK: Series have names and the logic for naming fields/series in a UDF is to take the first
-        # Expression's name. Note that this logic is tied to the `to_field` implementation of the Rust PythonUDF
-        # and is quite error prone! If our Series naming logic here is wrong, things will break when the UDF is run on a table.
-        name = evaluated_expressions[0].name()
-
-        # Post-processing of results into a Series of the appropriate dtype
-        if isinstance(result, Series):
-            return result.rename(name).cast(self.udf.return_dtype)._series
-        elif isinstance(result, list):
-            if self.udf.return_dtype == DataType.python():
-                return Series.from_pylist(result, name=name, pyobj="force")._series
-            else:
-                return Series.from_pylist(result, name=name, pyobj="allow").cast(self.udf.return_dtype)._series
-        elif _NUMPY_AVAILABLE and isinstance(result, np.ndarray):
-            return Series.from_numpy(result, name=name).cast(self.udf.return_dtype)._series
-        elif _PYARROW_AVAILABLE and isinstance(result, (pa.Array, pa.ChunkedArray)):
-            return Series.from_arrow(result, name=name).cast(self.udf.return_dtype)._series
-        else:
-            raise NotImplementedError(f"Return type not supported for UDF: {type(result)}")
```

Comment:
Moved this code into `run_udf` to allow us to inject the `func` from the caller, instead of creating it as part of the invocation,